core/mount/mount_linux.go

```diff
@@ -250,40 +249,72 @@ func getUnprivilegedMountFlags(path string) (int, error) {
 }
 
 func doPrepareIDMappedOverlay(tempRemountsLocation string, lowerDirs []string, usernsFd int) ([]string, func(), error) {
-	tmpLowerDirs := make([]string, 0, len(lowerDirs))
+	commonDir, err := getCommonDirectory(lowerDirs)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to determine common parent: %w", err)
+	}
 
 	cleanUp := func() {
-		for _, lowerDir := range tmpLowerDirs {
-			if err := unix.Unmount(lowerDir, 0); err != nil {
-				log.L.WithError(err).Warnf("failed to unmount temp lowerdir %s", lowerDir)
-				continue
-			}
-			// Using os.Remove() so if it's not empty, we don't delete files in the
-			// rootfs.
-			if err := os.Remove(lowerDir); err != nil {
-				log.L.WithError(err).Warnf("failed to remove temporary overlay lowerdir")
-			}
+		if err := unix.Unmount(tempRemountsLocation, 0); err != nil {
+			log.L.WithError(err).Warnf("failed to unmount idmapped directory %s", tempRemountsLocation)
 		}
-
-		// This dir should be empty now. Otherwise, we don't do anything.
+		// Using os.Remove() so if it's not empty, we don't delete files in the rootfs.
 		if err := os.Remove(tempRemountsLocation); err != nil {
-			log.L.WithError(err).Infof("failed to remove temporary overlay dir")
+			log.L.WithError(err).Infof("failed to remove idmapped directory")
 		}
 	}
-	for i, lowerDir := range lowerDirs {
-		tmpLowerDir := filepath.Join(tempRemountsLocation, strconv.Itoa(i))
-		tmpLowerDirs = append(tmpLowerDirs, tmpLowerDir)
 
-		if err := os.MkdirAll(tmpLowerDir, 0700); err != nil {
-			return nil, cleanUp, fmt.Errorf("failed to create temporary dir: %w", err)
-		}
-		if err := IDMapMountWithAttrs(lowerDir, tmpLowerDir, usernsFd, unix.MOUNT_ATTR_RDONLY, 0); err != nil {
-			return nil, cleanUp, err
-		}
+	// IDMapMount the directory containing all the layers
+	if err := IDMapMountWithAttrs(commonDir, tempRemountsLocation, usernsFd, unix.MOUNT_ATTR_RDONLY, 0); err != nil {
+		return nil, cleanUp, err
 	}
+
+	// Build new lower dir paths through the idmapped directory
+	tmpLowerDirs := buildIDMappedPaths(lowerDirs, commonDir, tempRemountsLocation)
+
 	return tmpLowerDirs, cleanUp, nil
 }
 
+// getCommonDirectory finds the common directory among the lowerDirs passed in.
+// "/" and "." are considered invalid common directories and are treated as error
+func getCommonDirectory(lowerDirs []string) (string, error) {
+	commonPrefix := longestCommonPrefix(lowerDirs)
+	if commonPrefix == "" {
+		return "", fmt.Errorf("no common prefix found")
+	}
+
+	// Ensure the common prefix ends at a directory boundary
+	if !strings.HasSuffix(commonPrefix, "/") {
+		lastSlash := strings.LastIndex(commonPrefix, "/")
+		if lastSlash == -1 {
+			return "", fmt.Errorf("no valid directory separator found in common prefix")
+		}
+		commonPrefix = commonPrefix[:lastSlash]
+	} else {
+		commonPrefix = strings.TrimSuffix(commonPrefix, "/")
```

Comment:
Why do we care to remove the last slash here? Can't we just avoid this else? I guess directories ending with "/" as a low dir should work, right?

Reply:
yeah, this was unnecessary. Updated this + tests to be ok about it!